src/app/_components/ListTruncator.js

```diff
@@ -0,0 +1,86 @@
+'use client';
+
+import React, { useCallback, useEffect, useRef, useState } from 'react';
+import P from 'prop-types';
+import { Show as ShowHint, Window as HintWindow } from './Hint';
+import { getRandomInt } from '@/utils/common';
+import { cn } from '@/utils/cn';
+
+export function ListTruncator({ id, content, hintContent, ellipsis, hintEllipsisClassName, hintWindowClassName }) {
+  const wrapperRef = useRef(null);
+  const containerRef = useRef(null);
+  const [overflown, setOverflown] = useState(false);
+
+  const calculateOverflow = useCallback(() => {
+    const wrapperWidth = wrapperRef.current.offsetWidth;
+    const containerWidth = containerRef.current.offsetWidth;
+    if (wrapperWidth > containerWidth) {
+      setOverflown(true);
+    } else {
+      setOverflown(false);
+    }
+  }, [wrapperRef, containerRef]);
+
+  useEffect(() => {
+    calculateOverflow();
+    window.addEventListener('resize', calculateOverflow);
+
+    return () => {
+      window.removeEventListener('resize', calculateOverflow);
+    };
+  }, [calculateOverflow]);
+
+  const hintId = `hint-for-${id}-${getRandomInt(0, 1000000)}`;
+
+  return (
+    <div ref={containerRef} className="max-w-full overflow-hidden">
+      <span ref={wrapperRef} className="relative inline-flex items-center gap-[8px]">
+        {content}
+      </span>
+
+      {overflown && (
+        <>
+          <ShowHint
+            opens={hintId}
+            actions={{
+              onMouseEnter: ({ open }) => open(hintId),
+              onMouseLeave: ({ close }) => close(),
+              onClick: ({ open }) => {
+                open(n => (n === hintId ? '' : hintId));
+              },
```

Comment:
I think this logic can be inside ShowHint component

Reply:
Thanks for comment! Moved this login into single `toggle` method.